### ISSUE TITLE
No Working Apple Silicon build going back as far as March

### ISSUE THREAD

**Issue by JudgeGroovyman** [bmx-ng/bmx-ng#189]
**Describe the bug**
The source code provided in the weekly builds does not build on Apple Silicon.

These same three errors are in the latest weekly build and also in a build from March 2024 (and likely from all builds in between):
```
[ 69%] Compiling:hook.bmx.release.macos.arm64.c
/Users/johnmcgarey/Downloads/BlitzMax/dist/bootstrap/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.c:18:26: error: incompatible function pointer types assigning to 'BBOBJECT (*)(BBINT, BBOBJECT, BBOBJECT)' (aka 'struct BBObject *(*)(int, struct BBObject *, struct BBObject *)') from 'void (*)()' [-Wincompatible-function-pointer-types]
   18 |         o->_brl_hook_thook_func = (&brl_blitz_NullFunctionError);
      |                                 ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.  

[ 71%] Compiling:threads.bmx.release.macos.arm64.c
/Users/johnmcgarey/Downloads/BlitzMax/dist/bootstrap/mod/brl.mod/threads.mod/.bmx/threads.bmx.release.macos.arm64.c:14:33: error: incompatible function pointer types assigning to 'BBOBJECT (*)(BBOBJECT)' (aka 'struct BBObject *(*)(struct BBObject *)') from 'void (*)()' [-Wincompatible-function-pointer-types]
   14 |         o->_brl_threads_tthread__entry = (&brl_blitz_NullFunctionError);
      |                                        ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

[ 85%] Compiling:bmk_pm_ng.bmx.release.macos.arm64.c
/Users/johnmcgarey/Downloads/BlitzMax/dist/bootstrap/src/bmk/.bmx/bmk_pm_ng.bmx.release.macos.arm64.c:113:40: error: incompatible function pointer types assigning to 'BBOBJECT (*)(BBOBJECT)' (aka 'struct BBObject *(*)(struct BBObject *)') from 'void (*)()' [-Wincompatible-function-pointer-types]
  113 |         o->__m_bmk_pm_ng_tthreadpooltask_func = (&brl_blitz_NullFunctionError);
      |                                               ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
That was of course just a listing of the errors in the full build output (both builds errors attached here):
[BlitzMax_macos_arm64_0.143.3.57.202403240147.error.txt](https://github.com/user-attachments/files/17373045/BlitzMax_macos_arm64_0.143.3.57.202403240147.error.txt)
[BlitzMax_macos_arm64_0.144.3.57.202410130219.error.txt](https://github.com/user-attachments/files/17373046/BlitzMax_macos_arm64_0.144.3.57.202410130219.error.txt)

**To Reproduce**
I followed the build instructions on by right clicking `run_me_first.command` and then clicking open and the build began and then I got the errors in the build output.

**Expected behavior**
1. It should build cleanly or 
2. Some workaround should be clearly published in the build notes or 
3. At the very least a fix that we could implement ourselves should be documented here.

**Version information (please complete the following information):**
 M1 2020 Macbook Air with MacOS Sonoma 14.5 and xcode and xcode-command-line-tools both updated yesterday.

These are the releases that did not work:
`BlitzMax_macos_arm64_0.144.3.57.202410130219`
`BlitzMax_macos_arm64_0.143.3.57.202403240147`

**Additional context**
It seems that other people have ultimately been reporting similar errors since May in [this issue 183](https://github.com/bmx-ng/bmx-ng/issues/183) but it has not received any notable attention or solutions or workarounds.

I wish I was smart enough about such things to figure this out myself but I'm looking forward to what someone smarter than me figures out.   Probably theres just some small code change I can make in 3 places that will fix it. 

Thanks!

P.S. HMU if you would like me to test builds or fixes or whatever.  I would like to help.

**Comment by sblank** [comment]
Ditto... still hoping for a fix.  I don't have the C expertise anymore to do much other than minor tweaking.  

Thanks!  Hoping to use BlitzMax again someday!

**Comment by sblank** [comment]
I'll add that I'll second JudgeGroovyman's offer.  I have been programming since punch cards, although it's been a couple of decades since I've written anything in C.  I can usually follow instructions, though, and would be happy to test builds, workarounds, etc. on my MacBook Pro M1 Max.

**Comment by sblank** [comment]
Nope... I was hoping.  Still getting the 'Incompatible function point types' error in 3 places during the bmk compilation.

[ 69%] Compiling:hook.bmx.release.macos.arm64.c
/Users/sblank/Desktop/BlitzMax/dist/bootstrap/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.c:18:26: error: incompatible function pointer types assigning to 'BBOBJECT (*)(BBINT, BBOBJECT, BBOBJECT)' (aka 'struct BBObject *(*)(int, struct BBObject *, struct BBObject *)') from 'void (*)()' [-Wincompatible-function-pointer-types]
   18 |         o->_brl_hook_thook_func = (&brl_blitz_NullFunctionError);
      |                                 ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

[ 76%] Compiling:threads.bmx.release.macos.arm64.c
/Users/sblank/Desktop/BlitzMax/dist/bootstrap/mod/brl.mod/threads.mod/.bmx/threads.bmx.release.macos.arm64.c:15:33: error: incompatible function pointer types assigning to 'BBOBJECT (*)(BBOBJECT)' (aka 'struct BBObject *(*)(struct BBObject *)') from 'void (*)()' [-Wincompatible-function-pointer-types]
   15 |         o->_brl_threads_tthread__entry = (&brl_blitz_NullFunctionError);
      |                                        ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

[ 85%] Compiling:bmk_pm_ng.bmx.release.macos.arm64.c
/Users/sblank/Desktop/BlitzMax/dist/bootstrap/src/bmk/.bmx/bmk_pm_ng.bmx.release.macos.arm64.c:113:40: error: incompatible function pointer types assigning to 'BBOBJECT (*)(BBOBJECT)' (aka 'struct BBObject *(*)(struct BBObject *)') from 'void (*)()' [-Wincompatible-function-pointer-types]
  113 |         o->__m_bmk_pm_ng_tthreadpooltask_func = (&brl_blitz_NullFunctionError);
      |                                               ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

[ 90%] Archiving:time.release.macos.arm64.a
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: can't open file: /Users/sblank/Desktop/BlitzMax/dist/bootstrap/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.o (No such file or directory)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: can't open file: /Users/sblank/Desktop/BlitzMax/dist/bootstrap/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.o (No such file or directory)

[ 94%] Archiving:io.release.macos.arm64.a
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: can't open file: /Users/sblank/Desktop/BlitzMax/dist/bootstrap/mod/brl.mod/threads.mod/.bmx/threads.bmx.release.macos.arm64.o (No such file or directory)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: can't open file: /Users/sblank/Desktop/BlitzMax/dist/bootstrap/mod/brl.mod/threads.mod/.bmx/threads.bmx.release.macos.arm64.o (No such file or directory)

Followed by errors indicating that the file can't be opened... I assume it mean 'bmk'

the 'bcc' binary appears to compile without a problem.

**Comment by GWRon** [comment]
Try to adopt your stuff as I did for linux and newer gcc versions:
bmx-ng/brl.mod#332

Maybe that fixes some issues for you.

**Comment by sblank** [comment]
Thanks for commenting GWRon... I tried modifying the make.bmk file (I found 3 of them) as you suggested, but the same errors were generated during the compilation on my M1 Max.

I do appreciate the suggestion, though!

**Comment by GWRon** [comment]
did you verify it appends the "-w" parameters to the build command? Asking as each OS has their own "path" in the make.bmk.

```
@define compileC
	local src = nvl(arg1, %infile%)
	local obj = nvl(arg2, %outfile%)
	local opts = arg3

	#opts = opts .. " " .. %cc_opts% .. " " .. %BMK_CC_OPTS%

	local ext = sys.ExtractExt(src)

	if bmk.Platform() == "macos" or bmk.Platform() == "osx" or bmk.Platform() == "ios" then
...
	cmd = cmd .. opts .. cc_opts .. %mod_ccopts% .. genDebug .. " -o " .. bmk.Quote(obj) .. " " .. bmk.Quote(src)

	if bmk.MultiSys( cmd, src, obj, "" ) ~= 0 then
		bmk.ThrowNew("Build Error: failed to compile " .. src)
	end

@end
```
You could simply prepend that line before that "cmd = cmd..."

```
	#only usable for C/ObjC but not C++.
	if ext == "c" then 
	  opts = opts .. " -Wno-incompatible-pointer-types -Wno-implicit-function-declaration -Wno-int-conversion"
	end
	cmd = cmd .. opts .. cc_opts .. %mod_ccopts% .. genDebug .. " -o " .. bmk.Quote(obj) .. " " .. bmk.Quote(src)
```

If you then start MaxIDE and enable "verbose":
![image](https://github.com/user-attachments/assets/66d84097-379a-4d97-94d8-4fb8ffa6b36d)
MaxIDE will spit out what params it adds to the compilation process.

Now theck if it calls gcc (or whatever it calls there) with these 3 additional parameters.

**Comment by sblank** [comment]
Thanks for the reply... yes, I actually placed the -w parameters in 
several different places in all make.bmk files (I found 3 make.bmk 
files).  I have never been able to get the MaxIDE to work at any time on 
the Max M1.  When I was able to use BlitzMax several months ago, I used 
VSCode as the IDE because MaxIDE was unusable.

I'll give this a fresh try today.

Thanks again!

On 2024-10-21 02:31, Ronny Otto wrote:

> did you verify it appends the "-w" parameters to the build command? 
> Asking as each OS has their own "path" in the make.bmk.
> 
> @define compileC
> local src = nvl(arg1, %infile%)
> local obj = nvl(arg2, %outfile%)
> local opts = arg3
> 
> #opts = opts .. " " .. %cc_opts% .. " " .. %BMK_CC_OPTS%
> 
> local ext = sys.ExtractExt(src)
> 
> if bmk.Platform() == "macos" or bmk.Platform() == "osx" or 
> bmk.Platform() == "ios" then
> ...
> cmd = cmd .. opts .. cc_opts .. %mod_ccopts% .. genDebug .. " -o " .. 
> bmk.Quote(obj) .. " " .. bmk.Quote(src)
> 
> if bmk.MultiSys( cmd, src, obj, "" ) ~= 0 then
> bmk.ThrowNew("Build Error: failed to compile " .. src)
> end
> 
> @end
> 
> You could simply prepend that line before that "cmd = cmd..."
> 
> #only usable for C/ObjC but not C++.
> if ext == "c" then
> opts = opts .. " -Wno-incompatible-pointer-types 
> -Wno-implicit-function-declaration -Wno-int-conversion"
> end
> cmd = cmd .. opts .. cc_opts .. %mod_ccopts% .. genDebug .. " -o " .. 
> bmk.Quote(obj) .. " " .. bmk.Quote(src)
> 
> If you then start MaxIDE and enable "verbose":
> image.png (view on web) [1]
> MaxIDE will spit out what params it adds to the compilation process.
> 
> Now theck if it calls gcc (or whatever it calls there) with these 3 
> additional parameters.
> 
> --
> Reply to this email directly, view it on GitHub [2], or unsubscribe 
> [3].
> You are receiving this because you commented.Message ID: 
> ***@***.***>


Links:
------
[1] 
https://github.com/user-attachments/assets/66d84097-379a-4d97-94d8-4fb8ffa6b36d
[2] https://github.com/bmx-ng/bmx-ng/issues/189#issuecomment-2425847876
[3]

**Comment by JudgeGroovyman** [comment]
### I got one of the files with errors to compile!
Progress!  

I have not applied this to the whole build yet but for just one of the files with errors either of these compiler options worked:

```
-Wno-error=incompatible-function-pointer-types 
or
-Wno-incompatible-function-pointer-types 
```

### Examples
This compiles: 
```
gcc -Wno-error=incompatible-function-pointer-types -arch arm64  -I$BMX_ROOT/mod/brl.mod/hook.mod -I$BMX_ROOT/mod -I$BMX_ROOT/mod/brl.mod/blitz.mod  -DNDEBUG -DTHREADED  -w -DBMX_NG -mmacosx-version-min=10.15 -fno-exceptions -c -O3 -s  -o $BMX_ROOT/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.o   $BMX_ROOT/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.c 
```

This also compiles:
```
gcc -Wno-incompatible-function-pointer-types  -arch arm64  -I$BMX_ROOT/mod/brl.mod/hook.mod -I$BMX_ROOT/mod -I$BMX_ROOT/mod/brl.mod/blitz.mod  -DNDEBUG -DTHREADED  -w -DBMX_NG -mmacosx-version-min=10.15 -fno-exceptions -c -O3 -s  -o $BMX_ROOT/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.o  $BMX_ROOT/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.c 

```

This (the original) does not compile:
```
gcc -arch arm64  -I$BMX_ROOT/mod/brl.mod/hook.mod -I$BMX_ROOT/mod -I$BMX_ROOT/mod/brl.mod/blitz.mod  -DNDEBUG -DTHREADED  -w -DBMX_NG -mmacosx-version-min=10.15 -fno-exceptions -c -O3 -s  -o $BMX_ROOT/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.o  $BMX_ROOT/mod/brl.mod/hook.mod/.bmx/hook.bmx.release.macos.arm64.c 
```


### How I found it?
I went into `/Users/me/tmp/BlitzMax/dist/bootstrap/src/bmk` and narrowed down the build script there so it just recompiles one files that has the error and I was able to try a bunch of options until one finally compiled.


### Why did this happen?
I suspect this:  
My gcc is clang16 and here it talks about how in clang16 incompatible pointer types were upgraded from a warning to an error which explains why this changed (probably last years version of clang considered it a warning and proceeded)
https://www.redhat.com/en/blog/new-warnings-and-errors-clang-16

**Comment by sblank** [comment]
I think you are correct about the clang issue.  I am currently using clang 18.1.8 and my initial thought was to downgrade to a previous version <16, but then my fear was that I would be breaking something that is currently operating well.

I haven't had the time to dig into placing the "-Wno-error=incompatible-function-pointer-types" option anywhere other than in the make.bmk files, but that didn't work for me at all.  I tried to place them logically, but...

Perhaps a pragma with the -Wno... in the offending files?  Just a thought.

**Comment by JudgeGroovyman** [comment]
### More Progress!
TLDR I finally got a maxide.app unfortunately it crashes instantly.  The little rectangle splash screen pops up for <1s and then a few seconds later the mac crash screen shows up.

Here are the crash logs I got.  What do these mean?  

[BlitzmaxCrashlog1.txt](https://github.com/user-attachments/files/17528362/BlitzmaxCrashlog1.txt)
[BlitzmaxCrashlog2.txt](https://github.com/user-attachments/files/17528361/BlitzmaxCrashlog2.txt)

**What should I try next?**



### What I did step by step

Note: I am aware that there are probably better ways of doing everything I did but I couldn't figure them out :)

@sblank if you want to follow along try this and ask any questions you have.  

From a virgin .zip file:
#### Adjust the .build file with the flag
1. run `./build_dist.sh` and copy paste the filenames that have the errors (shortcut: `./build_dist.sh | grep error`)
2. go into this build file `dist/bootstrap/src/bmk/bmk.console.release.macos.arm64.build` and search for the filenames in there and manually add the flag  `-Wno-incompatible-function-pointer-types` to the gcc line for each of those files.  like this: `gcc -Wno-incompatible-function-pointer-types -arch arm64  -I$BMX_ROOT/mod -I$BM...` (thats just part of the line) 
3. Hint: in mine they were on lines 296, 334, and 384

#### Adjust the make.bmk files
1. find them all with `find . | grep make.bmk`
2. Find the mac CC_Opts (which is on line 495) and change that line to this `globals.SetOption("cc_opts", "osversion", "-mmacosx-version-min=10.15 -Wno-incompatible-function-pointer-types")`  
3. Hint: Only the one in bin seems to matter `./bin/make.bmk` but it seems smart to change them all

#### Build
1. run `./build_dist.sh` again and it should finish with just 4 warnings about using a library twice `ld: warning: ignoring duplicate libraries: '-lSystem'`.  There should finally be a MaxIDE.app!

#### cleanup
1. finally `xattr -cr MaxIDE.app` if you don't do that then mac says it is damaged (this is pretty common with building on mac in my experience)

**Comment by sblank** [comment]
Nice!  I followed your instructions and had the same results as you indicated.  MaxIDE crashes (even after using xattr), but it has never worked for me.  

I tried compiling the teapot_text.bmx sample file and was able to get the build started, but it failed with the following error(s):

/Users/sblank/Desktop/BlitzMax/mod/sdl.mod/sdl.mod/SDL/src/thread/SDL_thread.c:294:41: error: incompatible integer to pointer conversion passing 'SDL_threadID' (aka 'unsigned long') to parameter of type 'void *' [-Wint-conversion]
294 |         BBThread * bbThread = bbThreadRegister(SDL_ThreadID());
|                                                ^~~~~~~~~~~~~~
/Users/sblank/Desktop/BlitzMax/mod/brl.mod/blitz.mod/blitz_thread.h:140:36: note: passing argument to parameter 'thd' here
140 | BBThread *bbThreadRegister( void * thd );
|                                    ^
1 error generated.
Build Error: failed to compile (256) /Users/sblank/Desktop/BlitzMax/mod/sdl.mod/sdl.mod/SDL/src/thread/SDL_thread.c


If time permits, I'll look into this a bit tomorrow.  However, things are looking up.... nice job JudgeGroovyman!

**Comment by sblank** [comment]
I think this is another recent clang issue and perhaps something like "-Wno-int-conversion" would suppress the error.  Now I have to figure out where to put the -Wno...

**Comment by sblank** [comment]
Bingo and booyah!!

I went into the BlitzMax/bin folder and added "-Wno-int-conversion" to line 495 of the make.bmk file as you specified in your instructions above:

2). Find the mac CC_Opts (which is on line 495) and change that line to this globals.SetOption("cc_opts", "osversion", "-mmacosx-version-min=10.15 -Wno-incompatible-function-pointer-types")

Then tried to build and run the teapot_text.bmx file and... wait for it... it worked!  I now have a beautiful rotating red teapot on my Mac M1 screen!

I'm using VSCode with the BlitzMax extension.

Thank you JudgeGroovyman... you da' man!

Stan

**Comment by sblank** [comment]
Now to see about MaxIde...  :-)

**Comment by JudgeGroovyman** [comment]
> Then tried to build and run the teapot_text.bmx file and... wait for it... it worked! I now have a beautiful rotating red teapot on my Mac M1 screen!

This is amazing way to go!

I don't know how to proceed from here but FYI I had maxide running on this M1 Mac sometime last year and iirc I built it with no errors back then.

**Comment by GWRon** [comment]
You were able to build it because you used an older clang version that time...

Similar to gcc 14 you need to pass these flags for now (until Brucey decided to fix the underlying issues properly).

**Comment by JudgeGroovyman** [comment]
That sounds right and I have no problem using those flags.

the only problem now is even with the flags the maxide crashes

**Comment by GWRon** [comment]
If you compile maxide as debug...what does the debugger tell you?

If you are not enjoying the command line (zhe debugger would ask you for key codes to navigate) you might run it via the vs code extension.

**Comment by sblank** [comment]
So far no luck with MaxIDE, but VSCode is working well.  I've even been able to get the raylib mod working by adding another -Wno statement (-Wno-implicit-function-declaration) to fix an error.  The raylib examples are working, with the exception of the Space Invaders game, which throws another '-Wincompatible-function-pointer-types' error... apparently the -Wno in make.bmk doesn't fix that one.  However, most of the other raylib examples are working.

I'm seeing the 'old' showing the graphics viewport in one-quarter of the window in some of the Raylib examples.  If memory serves, this was due to the original Retina screen resolution?  I was able to fix that in OpenGL at one point, but I haven't been able to locate the proper viewport command in BlitzMax yet.

**Comment by sblank** [comment]
Fixed the one-quarter screen resolution in the raylib.mod... still no luck with MaxIDE, though.

**Comment by JudgeGroovyman** [comment]
> If you compile maxide as debug...what does the debugger tell you?

@GWRon How do I run the debugger?  do you mean gdb?

After building with `-d -gdb` I tried gdb and got this error

`"/Users/johnmcgarey/tmp/BlitzMax/./MaxIDE.app/Contents/MacOS/maxide": not in executable format: file format not recognized`

![GDB Output](https://github.com/user-attachments/assets/4e88b8d9-bcbb-4c45-b6f4-d0aa9157b545)

**Comment by JudgeGroovyman** [comment]
Ok thank you for that information about the debugger that makes sense.

> Or you run the whole thing from a command line and might be able to see that it expects you to enter something as soon as it crashes (eg enter "h" there...).

When I did this I got this error and some stuff about fonts being replaced (the full 2 page listing and call stack is attached)
`*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSToolbarItem isEqualToString:]: unrecognized selector sent to instance 0x600002d06050'` 

[open MaxIDE.app error.txt](https://github.com/user-attachments/files/17560238/open.MaxIDE.app.error.txt)

**Comment by sblank** [comment]
The one-quarter screen resolution issue is fixed by multiplying the viewport/screen height and width x2.  For example, in the raylib shaders_julia_set.bmx example file, here is what I did:

' Create a RenderTexture2D to be used for render to texture
Local target:RRenderTexture2D = LoadRenderTexture(screenWidth*2, screenHeight*2)

and multiplying screenWidth and screenHeight by 2 fixed the issue.  Each program using OpenGL might have a different statement specifying the screen viewport dimensions and you would need to adjust each one accordingly.  Here is another example in the shaders_raymarching.bmx file that displays a set of geometric solids and a gray checkerboard scene:

Local resolution:Float[] = [Float(screenWidth*2), Float(screenHeight*2)]

Again, finding the correct statement that sets the viewport and muliplying width and height x2 seems to work well.

**Comment by sblank** [comment]
I have guile installed (I have a thing for scheme and lisp... not a guru, but I'm fascinated by the lisp family).

I'll take a look at the debug information and see if my old brain can wake up a bit... well, I have the right hope, anyway :-)

**Comment by sblank** [comment]
I had to edit the quarter-screen code above to include the '*' sign in the statement... for some reason the original markup omitted the asterisk.

**Comment by GWRon** [comment]
@sblank ...wrap code in three back ticks... \`\`\`

**Comment by sblank** [comment]
Thanks @GWRon... that's a great tip!

**Comment by sblank** [comment]
The latest Mac arm build seems to be working properly with the exception of MaxIDE.  I was able to compile the bcc and bmk binaries and run some sample programs.  I had to add '-Wno-incompatible-function-pointer-types' to line 495 in the make.bmk file to avoid an error in compiling MaxIDE, but that didn't seems to help.

However, very pleased with the progress and I enjoy using BlitzMax again!  Thank you woollybah, JudgeGroovyman, and GWRon!
